compute/cker/include/cker/train/operation/MaxPool.h

```diff
@@ -0,0 +1,131 @@
+// https://github.com/Samsung/ONE/blob/a380292/compute/cker/include/cker/operation/MaxPool.h#L42-L88
+// 'arg_max_index' is to record max-arguments' index to apply gradient later.
+inline void MaxPool2D(const PoolParams &params, const Shape &input_shape, const float *input_data,
+                      const Shape &output_shape, float *output_data, int *arg_max_index)
+{
+  assert(input_shape.DimensionsCount() == 4);
+  assert(output_shape.DimensionsCount() == 4);
+  assert(input_shape.Dims(3) == output_shape.Dims(3)); // MaxPool2D doesn't change depth
+
+  const int batches = MatchingDim(input_shape, 0, output_shape, 0);
+  const int input_height = input_shape.Dims(1);
+  const int input_width = input_shape.Dims(2);
+  const int output_height = output_shape.Dims(1);
+  const int output_width = output_shape.Dims(2);
+  const int stride_height = params.stride_height;
+  const int stride_width = params.stride_width;
+
+  const auto in_mat = MapAsMatrixWithLastDimAsRows(input_data, input_shape);
+  auto out_mat = MapAsMatrixWithLastDimAsRows(output_data, output_shape);
+  auto arg_max_index_mat = MapAsMatrixWithLastDimAsRows(arg_max_index, output_shape);
+
+  // Prefill the output to minimum representable float value
+  out_mat.setConstant(std::numeric_limits<float>::lowest());
+
+  for (int b = 0; b < batches; ++b)
+  {
+    for (int h = 0; h < input_height; ++h)
+    {
+      for (int w = 0; w < input_width; ++w)
+      {
+        // (h_start, h_end) * (w_start, w_end) is the range that the input
+        // vector projects to.
+        int hpad = h + params.padding_values.height;
+        int wpad = w + params.padding_values.width;
+
+        int h_start =
+          (hpad < params.filter_height) ? 0 : (hpad - params.filter_height) / stride_height + 1;
+        int h_end = std::min(hpad / stride_height + 1, output_height);
+
+        int w_start =
+          (wpad < params.filter_width) ? 0 : (wpad - params.filter_width) / stride_width + 1;
+        int w_end = std::min(wpad / stride_width + 1, output_width);
+
+        // compute elementwise sum
+        for (int ph = h_start; ph < h_end; ++ph)
+        {
+          for (int pw = w_start; pw < w_end; ++pw)
+          {
+            const int out_offset = NodeOffset(b, ph, pw, output_height, output_width);
+            const int in_offset = NodeOffset(b, h, w, input_height, input_width);
+
+            const auto out_vector = out_mat.col(out_offset);
+            const auto in_vector = in_mat.col(in_offset);
+
+            // update arg_max_index_mat
+            arg_max_index_mat.col(out_offset) =
+              (out_vector.array() < in_vector.array())
+                .select(in_offset, arg_max_index_mat.col(out_offset));
+
+            // update out_mat
+            out_mat.col(out_offset) = out_vector.cwiseMax(in_vector);
+          }
+        }
+      }
+    }
+  }
+}
+
+inline void MaxPool2DGrad(const Shape &deriv_output_shape, const float *deriv_output_data,
```

Comment:
Additionally also here, `MaxPool2DGrad` only do maxpool thing, not activation related. 
So, To make corresponded `MaxPool2D`, I thought it looks more natural that `MaxPool2D` only does max pooling, not activation.

Reply:
@zetwhite We handles several types (e.g. max, avg, ...) of pooling in `PoolLayer`. It handles activation (including no activation) using `*_activation_min/max` in parameter. It is the way how `tflite` handles activation. We used similar (or same) kernel if there is no reason to do differently. Thus, handling differently only in `MaxPool`'s `forward` may make others confused. I prefer to keep the same style.

Reply:
@glistening Oh, I fully understood your intention. I'll move activation part to cker.

Reply:
updated!